packages/components/src/panel/body.js

```diff
@@ -2,94 +2,139 @@
  * External dependencies
  */
 import classnames from 'classnames';
+import { noop } from 'lodash';
+import mergeRefs from 'react-merge-refs';
+import {
+	useDisclosureState,
+	Disclosure,
+	DisclosureContent,
+} from 'reakit/Disclosure';
 
 /**
  * WordPress dependencies
  */
-import { Component, forwardRef } from '@wordpress/element';
+import { useReducedMotion } from '@wordpress/compose';
+import { forwardRef, useRef } from '@wordpress/element';
 import { chevronUp, chevronDown } from '@wordpress/icons';
 
 /**
  * Internal dependencies
  */
 import Button from '../button';
 import Icon from '../icon';
+import { useUpdateEffect } from '../utils';
 
-export class PanelBody extends Component {
-	constructor( props ) {
-		super( ...arguments );
-		this.state = {
-			opened: props.initialOpen === undefined ? true : props.initialOpen,
-		};
-		this.toggle = this.toggle.bind( this );
+export function PanelBody(
+	{
+		children,
+		className,
+		disableSmoothScrollIntoView,
+		focusable,
+		icon,
+		initialOpen: initialOpenProp,
+		onToggle = noop,
+		opened,
+		title,
+	},
+	ref
+) {
+	const initialOpen = useRef( initialOpenProp ).current;
+	const disclosure = useDisclosureState( {
+		visible: initialOpen !== undefined ? initialOpen : opened,
+	} );
+	const nodeRef = useRef();
+
+	// Defaults to 'smooth' scrolling
+	// https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView
+	let scrollBehavior = useReducedMotion() ? 'auto' : 'smooth';
+
+	// However, this behavior can be overridden by prop
+	if ( disableSmoothScrollIntoView ) {
+		scrollBehavior = 'auto';
 	}
 
-	toggle( event ) {
-		event.preventDefault();
-		if ( this.props.opened === undefined ) {
-			this.setState( ( state ) => ( {
-				opened: ! state.opened,
-			} ) );
-		}
+	const isOpened = disclosure.visible;
 
-		if ( this.props.onToggle ) {
-			this.props.onToggle();
+	// Runs after initial render
+	useUpdateEffect( () => {
+		if ( disclosure.visible ) {
+			/*
+			 * Scrolls the content into view when visible.
+			 * This improves the UX when there are multiple stacking <PanelBody />
+			 * components in a scrollable container.
+			 */
+			if ( nodeRef.current.scrollIntoView ) {
+				nodeRef.current.scrollIntoView( {
+					inline: 'nearest',
+					block: 'nearest',
+					behavior: scrollBehavior,
+				} );
+			}
 		}
-	}
 
-	render() {
-		const {
-			title,
-			children,
-			opened,
-			className,
-			icon,
-			forwardedRef,
-		} = this.props;
-		const isOpened = opened === undefined ? this.state.opened : opened;
-		const classes = classnames( 'components-panel__body', className, {
-			'is-opened': isOpened,
-		} );
+		onToggle( disclosure.visible );
+	}, [ disclosure.visible, onToggle, scrollBehavior ] );
+
+	useUpdateEffect( () => {
+		disclosure.setVisible( opened );
+	}, [ disclosure.setVisible, opened ] );
```

Comment:
For me using effects to sync between disclosure state and external state doesn't seems like a great API. It seems thought l like we're forced to do it this way because of how Reakit disclosure API is modeled. I wonder personally if the Reakit Disclosure API could be made more friendly for "controlled components". cc  @diegohaz

Reply:
`useDisclosureState` (and other Reakit state hooks) is analogous to `React.useState`. The argument you pass in is the initial state, and you would have to call `setState` to update it from an external prop:

If you use `React.useState` inside a component, but you want to make this component optionally controlled, you would have to do this:
```jsx
function OptionallyControlledComponent(props) {
  const [visible, setVisible] = React.useState(props.visible);

  if (props.visible !== undefined && props.visible !== visible) {
    // React will bail out from rendering since this is called in the render phase
    setVisible(props.visible);
  }

  ...
}
```

This is basically `getDerivedStateFromProps` in function components. You can do the same with `useDisclosureState`, there's no need for `useEffect`:
```jsx
function OptionallyControlledComponent(props) {
  const disclosure = useDisclosureState({ visible: props.visible });
  
  if (props.visible !== undefined && props.visible !== disclosure.visible) {
    disclosure.setVisible(props.visible);
  }
  
  return (
    <>
      <Disclosure {...disclosure}>Toggle</Disclosure>
      <DisclosureContent {...disclosure}>Content</DisclosureContent>
    </>
  );
}
```

In both cases above you would be able to pass a `visible` prop to have a controlled component:
```jsx
<OptionallyControlledComponent visible />
```
Or pass nothing, and this would be an uncontrolled component:
```jsx
<OptionallyControlledComponent />
```
You can even add a `defaultVisible` prop that would be passed to `useState`/`useDisclosureState` so it works exactly as a React uncontrolled form element (with `value`/`defaultValue`):
```jsx
<OptionallyControlledComponent defaultVisible />
```

But you can always opt out from using the state hook inside the component and make it fully controlled by passing the required state props to the components:

```jsx
function ControlledComponent(props) {
  const baseId = useInstanceId();
  return (
    <>
      <Disclosure
        visible={props.visible}
        baseId={baseId}
        toggle={props.onToggle}
      >
        Toggle
      </Disclosure>
      <DisclosureContent baseId={baseId} visible={props.visible}>Content</DisclosureContent>
    </>
  );
}
```

You can find the list of state props [here](https://reakit.io/docs/disclosure/#disclosure):

![Screenshot of the page on the link above](https://user-images.githubusercontent.com/3068563/85594696-88494a00-b61e-11ea-98d9-e454cd1d9944.png)

Unfortunately, it's still not showing which props are required and optional. This is something we should improve in the documentation. But, if the editor supports TypeScript (like VSCode), you'll see the required props without the `?` symbol.

![Screenshot of VSCode showing the required props from the Disclosure component in an autocomplete dropdown](https://user-images.githubusercontent.com/3068563/85595908-a5cae380-b61f-11ea-893e-ac1b16127ef1.png)

Here's a CodeSandbox with some examples: https://codesandbox.io/s/controlled-and-uncontrolled-reakit-components-rgwck

That said, it's not all set in stone. There's an open issue about this (https://github.com/reakit/reakit/issues/487). Because of the consistency with `React.useState`, the discussion there is leaning more towards maintaining the current API.

Reply:
@diegohaz I just tried out the `OptionallyControlledComponent` implementation. It resulted in some undesired recursive action 🙈 

![Screen Capture on 2020-06-24 at 13-21-28](https://user-images.githubusercontent.com/2322354/85602906-d6118280-b61d-11ea-880d-c88daa98f58d.gif)

(The GIF doesn't fully capture it. It was flipping REALLY fast)

The component seems to be pretty happy with the `useUpdatedEffect` wrapper:

```js
	useUpdateEffect( () => {
		disclosure.setVisible( opened );
	}, [ disclosure.setVisible, opened ] );
```

I hear where @youknowriad is coming from.
I've noticed and felt the challenge of creating an optimal and seamless "Optionally Controlled Component" experience where the consumer doesn't have to do too much to manage state and props.

With that said, I'm okay with the current implementation, with the idea of smoothening it out later.

I think my upcoming updates for the `useControlledState` hook may help?
https://github.com/WordPress/gutenberg/blob/d510d27c32bbdf3fade4c4317d38ddd5f04dcf68/packages/components/src/utils/hooks/use-controlled-state.js

That being part of this PR:
https://github.com/WordPress/gutenberg/pull/23006

Reply:
Just to clarify, i'm ok with the current implementation too but I do believe both useState and useDisclosureState favor local state where in our components (and I believe most UI libraries but I'm not sure), we don't build "optinally controlled components", I feel we default to "controlled components" and the uncontrolled ones are the exception.

In pure React, you  don't need useState and effects to achieve that you just pass the props down.

Reply:
I believe it's because you're calling `onToggle` whenever `disclosure.visible` changes, which changes `opened`, which then changes `disclosure.visible` again and this keeps happening...